electrum/wizard.py

```diff
@@ -351,18 +408,102 @@ def keystore_from_data(self, wallet_type, data):
+        if not self.is_multisig(wizard_data):
+            return True, ''
+
+        # current cosigner might be incomplete. In that case, return valid
+        cosigner_data = self.current_cosigner(wizard_data)
+        if self.needs_derivation_path(wizard_data):
+            if 'derivation_path' not in cosigner_data:
+                self.logger.debug('defer multisig check: missing derivation_path')
+                return True, ''
+        if self.wants_ext(wizard_data):
+            if 'seed_extra_words' not in cosigner_data:
+                self.logger.debug('defer multisig check: missing extra words')
+                return True, ''
+        if self.is_current_cosigner_hardware(wizard_data):
+            if 'master_key' not in cosigner_data:
+                self._logger.debug('defer multisig check: missing master_key')
+                return True, ''
+
+        user_info = ''
+
+        if self.has_duplicate_masterkeys(wizard_data):
+            self._logger.debug('Duplicate master keys!')
+            user_info = _('Duplicate master keys')
+            multisig_keys_valid = False
+        elif self.has_heterogeneous_masterkeys(wizard_data):
+            self._logger.debug('Heterogenous master keys!')
+            user_info = _('Heterogenous master keys')
+            multisig_keys_valid = False
+        else:
+            multisig_keys_valid = True
+
+        return multisig_keys_valid, user_info
+
+    def validate_seed(self, seed: str, seed_variant: str, wallet_type: str):
+        seed_type = ''
+        seed_valid = False
+        validation_message = ''
+
+        if seed_variant == 'electrum':
+            seed_type = mnemonic.seed_type(seed)
+            if seed_type != '':
+                seed_valid = True
+        elif seed_variant == 'bip39':
+            is_checksum, is_wordlist = keystore.bip39_is_checksum_valid(seed)
+            status = ('checksum: ' + ('ok' if is_checksum else 'failed')) if is_wordlist else 'unknown wordlist'
+            validation_message = 'BIP39 (%s)' % status
+
+            if is_checksum:
+                seed_type = 'bip39'
+                seed_valid = True
+        elif seed_variant == 'slip39':
+            # seed shares should be already validated by wizard page, we have a combined encrypted seed
+            if seed and isinstance(seed, EncryptedSeed):
+                seed_valid = True
+                seed_type = 'slip39'
+            else:
+                seed_valid = False
+        else:
+            raise Exception(f'unknown seed variant {seed_variant}')
+
+        # check if seed matches wallet type
+        if wallet_type == '2fa' and not is_any_2fa_seed_type(seed_type):
+            seed_valid = False
+        elif wallet_type == 'standard' and seed_type not in ['old', 'standard', 'segwit', 'bip39', 'slip39']:
+            seed_valid = False
+        elif wallet_type == 'multisig' and seed_type not in ['standard', 'segwit', 'bip39']:
+            seed_valid = False
```

Comment:
why no `slip39` nesting in multisig? :P
I think the old wizard allows it.

Reply:
I'll have a look

Reply:
probably works already, and the storage creation code is just being conservative.